src/shogun/mathematics/linalg/LinalgBackendEigen.h

```diff
@@ -544,8 +553,8 @@ namespace shogun
 		T cross_entropy_impl(const SGMatrix<T>& p, const SGMatrix<T>& q) const;
 
 		/** Eigen3 vector dot-product method */
-		template <typename T>
-		T dot_impl(const SGVector<T>& a, const SGVector<T>& b) const;
+		template <typename T, typename U, typename TU = typename linalg::promote<T, U>::type>
```

Comment:
I wonder whether it would be a point to let the caller decide the return type? But maybe that is overkill ...

Reply:
it should be possible if you provided `TU`. However, it would throw an error since these casses aren't specialized.